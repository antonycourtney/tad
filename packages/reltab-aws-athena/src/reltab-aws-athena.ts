import * as log from "loglevel";
import {
  TableRep,
  QueryExp,
  Schema,
  DataSourceConnection,
  defaultEvalQueryOptions,
  EvalQueryOptions,
  DataSourceProvider,
  registerProvider,
  DbDriver,
  SQLDialect,
  DbDataSource,
  ColumnStatsMap,
} from "reltab";
import {
  LeafSchemaMap,
  Row,
  ColumnMetaMap,
  DataSourceId,
  PrestoDialect,
  DataSourceNode,
  DataSourcePath,
} from "reltab";
import * as aws from "aws-sdk";
import * as path from "path";

const AthenaExpress = require("athena-express");

const REGION = "us-west-2";

aws.config.update({ region: REGION });
const athenaExpressConfig = { aws }; //configuring athena-express with aws sdk object
const athenaExpress = new AthenaExpress(athenaExpressConfig);

const mapIdent = (src: string): string => {
  const ret = src.replace(/[^a-z0-9_]/gi, "_");
  return ret;
};

const isAlpha = (ch: string): boolean => /^[A-Z]$/i.test(ch);

export class AWSAthenaDriver implements DbDriver {
  readonly sourceId: DataSourceId;
  readonly dialect: SQLDialect = PrestoDialect;
  tableMap: LeafSchemaMap;

  constructor() {
    this.sourceId = {
      providerName: "aws-athena",
      resourceId: "",
    };
    this.tableMap = {};
  }

  async getDisplayName(): Promise<string> {
    return "AWS Athena";
  }

  async runSqlQuery(sqlQuery: string): Promise<Row[]> {
    const qres = await athenaExpress.query(sqlQuery);
    const rows = qres.Items as Row[];
    return rows;
  }

  async importCsv(): Promise<void> {
    throw new Error("importCsv not implemented for aws-athena");
  }

  async getTableSchema(tableName: string): Promise<Schema> {
    const sqlQuery = `DESCRIBE ${tableName}`;

    const qres = await athenaExpress.query(sqlQuery);
    const items = qres.Items; // each item has one key (column name), mapped to its type.

    const extendCMap = (cmm: ColumnMetaMap, item: any): ColumnMetaMap => {
      const cnm = Object.keys(item)[0] as string;
      const cType = (item[cnm] as string).toLocaleUpperCase();

      const cmd = {
        displayName: cnm,
        columnType: cType,
      };
      cmm[cnm] = cmd;
      return cmm;
    };

    const columnIds = items.map((item: any) => Object.keys(item)[0]);
    const cmMap = items.reduce(extendCMap, {});
    const schema = new Schema(PrestoDialect, columnIds, cmMap);
    return schema;
  }

  async getSqlQuerySchema(sqlQuery: string): Promise<Schema> {
    throw new Error(
      "AWSAthenaDriver.getSqlQuerySchema: base sql queries not supported"
    );
  }

  async getSqlQueryColumnStatsMap(sqlQuery: string): Promise<ColumnStatsMap> {
    return {};
  }

  async getRootNode(): Promise<DataSourceNode> {
    throw new Error("getRootNode not yet implemented for aws-athena");
  }

  async getChildren(dsPath: DataSourcePath): Promise<DataSourceNode[]> {
    // TODO
    return [];
  }
  async getTableName(dsPath: DataSourcePath): Promise<string> {
    const { path } = dsPath;
    if (path.length < 1) {
      throw new Error("getTableName: empty path");
    }
    return path[path.length - 1];
  }
}

const awsAthenaDataSourceProvider: DataSourceProvider = {
  providerName: "aws-athena",
  connect: async (resourceId: any): Promise<DataSourceConnection> => {
    const driver = new AWSAthenaDriver();
    const dsConn = new DbDataSource(driver);
    return dsConn;
  },
};

registerProvider(awsAthenaDataSourceProvider);
