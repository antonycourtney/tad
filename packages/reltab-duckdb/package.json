{
  "name": "reltab-duckdb",
  "version": "0.14.0",
  "description": "sqlite driver for reltab",
  "main": "dist/reltab-duckdb.js",
  "types": "dist/reltab-duckdb.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "clean": "rimraf dist",
    "build": "tsc -p tsconfig-build.json",
    "test": "AWS_PROFILE=dev ./scripts/run-tests-locally.sh",
    "test:run": "jest --config jest.config.json --no-cache"
  },
  "keywords": [
    "relational",
    "sql",
    "database",
    "sqlite"
  ],
  "author": "Antony Courtney <antony@antonycourtney.com>",
  "license": "MIT",
  "dependencies": {
    "duckdb-async": "^1.0.0",
    "loglevel": "^1.8.0",
    "pretty-hrtime": "^1.0.3"
  },
  "devDependencies": {
    "@types/jest": "^27.4.0",
    "@types/node": "^18.7.18",
    "jest": "^27.4.6",
    "mkdirp": "^1.0.4",
    "prettier": "^2.5.1",
    "reltab": "^0.12.0",
    "ts-jest": "^27.1.2",
    "typescript": "^4.8.3"
  },
  "peerDependencies": {
    "reltab": "^0.12.0"
  }
}
