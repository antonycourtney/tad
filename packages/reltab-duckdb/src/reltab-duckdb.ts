import { Connection, Database } from "duckdb-async";
import * as log from "loglevel";
import {
  colIsNumeric,
  ColumnMetadata,
  ColumnMetaMap,
  ColumnStatsMap,
  ColumnType,
  DataSourceConnection,
  DataSourceId,
  DataSourceNode,
  DataSourcePath,
  DataSourceProvider,
  DbDataSource,
  DbDriver,
  DuckDBDialect,
  NumericSummaryStats,
  registerProvider,
  Row,
  Schema,
  SQLDialect,
  TextSummaryStats,
} from "reltab"; // eslint-disable-line
import { initS3 } from "./s3utils";

export * from "./csvimport";

const columnTypes = DuckDBDialect.columnTypes;

let viewCounter = 0;

const genViewName = (): string => `tad_tmpView_${viewCounter++}`;

const typeLookup = (tnm: string): ColumnType => {
  const ret = columnTypes[tnm] as ColumnType | undefined;
  if (ret == null) {
    throw new Error("typeLookup: unknown type name: '" + tnm + "'");
  }
  return ret;
};

/* A little ConnectionPool class because it turns out node-duckdb
 * doesn't allow concurrent queries on a connection.
 */
class ConnectionPool {
  db: Database;
  private pool: Connection[];

  constructor(db: Database) {
    this.db = db;
    this.pool = [];
  }

  async take(): Promise<Connection> {
    if (this.pool.length > 0) {
      return this.pool.pop()!;
    } else {
      const conn = await this.db.connect();
      await initS3(conn);
      return conn;
    }
  }

  giveBack(conn: Connection) {
    this.pool.push(conn);
  }
}

const parsePercentage = (s: string | undefined): number | null => {
  if (s != undefined && s.endsWith("%")) {
    const noPct = s.replace(/%$/, "");
    const ret = Number.parseFloat(noPct) / 100.0;
    return ret;
  }
  return null;
};

/**
 * checkCols: check that all of the specified column names are present in the row and non-null
 */
function checkCols(row: Row, colNames: string[]): boolean {
  for (const colName of colNames) {
    if (row[colName] == null) {
      return false;
    }
  }
  return true;
}

/**
 * Take the rows from a table_info() or DESCRIBE and turn it into
 * a reltab Schema
 * @param metaRows
 */
export function schemaFromTableInfo(
  metaRows: Row[],
  columNameKey: string,
  columnTypeKey: string
): Schema {
  const extendCMap = (
    columnMetaMap: ColumnMetaMap,
    row: any,
    idx: number
  ): ColumnMetaMap => {
    const displayName = row[columNameKey];
    const columnType: string = row[columnTypeKey].toLocaleUpperCase();
    const ct = DuckDBDialect.columnTypes[columnType];
    const columnMetadata: ColumnMetadata = {
      displayName,
      columnType,
    };
    columnMetaMap[displayName] = columnMetadata;
    return columnMetaMap;
  };

  const cmMap = metaRows.reduce(extendCMap, {});
  const columnIds = metaRows.map((r) => r[columNameKey]);
  const schema = new Schema(DuckDBDialect, columnIds as string[], cmMap);
  return schema;
}

/**
 * Take the rows from a table_info() or DESCRIBE and turn it into
 * a reltab Schema
 * @param metaRows
 */
export function columnStatsFromSummarize(
  metaRows: Row[],
  columNameKey: string,
  columnTypeKey: string
): ColumnStatsMap {
  const columnStatsMap: ColumnStatsMap = {};

  for (const row of metaRows) {
    const colId = row[columNameKey] as string;
    const columnType: string = (
      row[columnTypeKey] as string
    ).toLocaleUpperCase();
    const ct = DuckDBDialect.columnTypes[columnType];
    if (
      ct &&
      colIsNumeric(ct) &&
      checkCols(row, [
        "min",
        "max",
        "approx_unique",
        "count",
        "null_percentage",
      ])
    ) {
      // numeric type!
      // annoyingly, DuckDb summarize stats returned as a (nullable!) varchar:
      const minVal = Number.parseFloat(row.min as string);
      const maxVal = Number.parseFloat(row.max as string);
      const approxUnique = Number.parseInt(row.approx_unique as string);
      const count = Number.parseInt(row.count as string);
      const pctNull = parsePercentage(row.null_percentage as string);
      const columnStats: NumericSummaryStats = {
        statsType: "numeric",
        min: minVal,
        max: maxVal,
        approxUnique,
        count,
        pctNull,
      };
      columnStatsMap[colId] = columnStats;
    }
  }
  return columnStatsMap;
}

export class DuckDBDriver implements DbDriver {
  readonly displayName: string;
  readonly sourceId: DataSourceId;
  readonly dialect: SQLDialect = DuckDBDialect;
  dbfile: string;
  db: Database;
  connPool: ConnectionPool;

  constructor(dbfile: string, db: Database, dbConn: Connection) {
    this.dbfile = dbfile;
    this.displayName = dbfile;
    this.sourceId = { providerName: "duckdb", resourceId: dbfile };
    this.db = db;
    this.connPool = new ConnectionPool(db);
  }

  async runSqlQuery(query: string): Promise<Row[]> {
    const conn = await this.connPool.take();
    let ret: any;
    try {
      log.info("runSqlQuery:\n", query);
      ret = await conn.all(query);
    } finally {
      this.connPool.giveBack(conn);
    }
    return ret;
  }

  async getDisplayName(): Promise<string> {
    return this.displayName;
  }

  async getTableSchema(tableName: string): Promise<Schema> {
    return this.getSqlQuerySchema(tableName);
  }

  async getSqlQuerySchema(sqlQuery: string): Promise<Schema> {
    let descRows: Row[];
    const describeQuery = `describe ${sqlQuery}`;
    descRows = await this.runSqlQuery(describeQuery);

    const schema = schemaFromTableInfo(descRows, "column_name", "column_type");
    return schema;
  }

  async getSqlQueryColumnStatsMap(sqlQuery: string): Promise<ColumnStatsMap> {
    try {
      const summarizeQuery = `summarize ${sqlQuery}`;
      const descRows = await this.runSqlQuery(summarizeQuery);
      const columnStatsMap = columnStatsFromSummarize(
        descRows,
        "column_name",
        "column_type"
      );
      return columnStatsMap;
    } catch (err) {
      console.warn("*** summarize query failed: ", err);
      return {};
    }
  }

  async getRootNode(): Promise<DataSourceNode> {
    const rootNode: DataSourceNode = {
      id: this.dbfile,
      kind: "Database",
      displayName: this.dbfile,
      isContainer: true,
    };
    return rootNode;
  }
  async getChildren(dsPath: DataSourcePath): Promise<DataSourceNode[]> {
    const { path } = dsPath;
    let node: DataSourceNode;
    const tiQuery = `PRAGMA show_tables;`;
    const dbRows = await this.runSqlQuery(tiQuery);
    const tableNames: string[] = dbRows.map((row: Row) => row.name as string);
    const childNodes: DataSourceNode[] = tableNames.map((tableName) => ({
      id: tableName,
      kind: "Table",
      displayName: tableName,
      isContainer: false,
    }));
    return childNodes;
  }

  async getTableName(dsPath: DataSourcePath): Promise<string> {
    const { path } = dsPath;
    if (path.length < 1) {
      throw new Error("getTableName: empty path");
    }
    return path[path.length - 1];
  }
}

const loadExtensions = async (db: Database): Promise<void> => {
  try {
    const ret = await db.exec(`INSTALL 'httpfs'; LOAD 'httpfs'`);
  } catch (err) {
    log.error("caught exception loading extensions: ", err);
    log.error("(ignoring unloadable extensions...)");
  }
};

const duckdbDataSourceProvider: DataSourceProvider = {
  providerName: "duckdb",
  connect: async (resourceId: any): Promise<DataSourceConnection> => {
    const dbfile = resourceId as string;
    const db = await Database.create(dbfile);
    await loadExtensions(db);
    // turn on fast parallel CSV loading:
    // await db.exec("SET experimental_parallel_csv=true;");
    const dbConn = await db.connect();
    const driver = new DuckDBDriver(dbfile, db, dbConn);
    const dsConn = new DbDataSource(driver);
    return dsConn;
  },
};

registerProvider(duckdbDataSourceProvider);
