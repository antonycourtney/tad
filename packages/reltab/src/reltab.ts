export * from "./dialectRegistry";
export * from "./dialect";
export * from "./remote/Connection";
export { getConnection } from "./remote/server";
export * from "./defs";
export * from "./histogram";
export * from "./FilterExp";
export * from "./QueryExp";
export * from "./ColumnStats";
export * from "./Schema";
export * from "./TableRep";
export * from "./AggFn";
export * from "./ColumnType";
export * from "./DataSource";
export * from "./toSql";
export { AggColSpec, ColumnMapInfo } from "./QueryRep";
export * from "./remote/Transport";
export * from "./remote/server";
export * from "./apiUtils";
