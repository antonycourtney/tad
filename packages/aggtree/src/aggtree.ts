import * as reltab from "reltab";
import * as _ from "lodash";
import { Path, PathTree } from "./PathTree";
import {
  DataSourceConnection,
  QueryExp,
  Schema,
  AggColSpec,
  asString,
  ValExp,
  ColumnMapInfo,
} from "reltab"; // eslint-disable-line
import { convertTypeAcquisitionFromJson } from "typescript";

export * from "./PathTree";

const { col, constVal } = reltab;
const PATHSEP = "#";
const ENCPATHSEP = "%23";
/*
 *  We used to use encodeURIComponent, but this isn't readily available on
 *  SQLite and we need to use the same encoding in both places so that
 *  path sort order works out correctly.
 *
 *  We'll do this much simpler string encoding that escapes % chars and PATHSEP.
 *  Can still be decoded with decodeURIComponent.
 */

const simpleStringEncode = (str?: string | null): string | undefined | null => {
  if (str == null) {
    return null;
  }

  return str.replace("%", "%25").replace(PATHSEP, ENCPATHSEP);
};

export const encodePath = (path: Path): string => {
  // const eps = path.map(encodeURIComponent)
  const eps = path.map(simpleStringEncode);
  const ret = PATHSEP + eps.join(PATHSEP);
  return ret;
};
export const decodePath = (pathStr: string): Path => {
  pathStr = pathStr.slice(1); // drop leading PATHSEP

  const eps = pathStr.length > 0 ? pathStr.split(PATHSEP) : [];
  const path = eps.map(decodeURIComponent);
  return path;
};

const addPathCols = (
  baseQuery: QueryExp,
  baseDepth: number,
  maxDepth: number
): QueryExp => {
  let retQuery = baseQuery;

  for (let i = baseDepth; i < maxDepth - 1; i++) {
    retQuery = retQuery.extend("_path" + i, asString(constVal(null)));
  }

  return retQuery;
};

export class VPivotTree {
  rt: DataSourceConnection;
  baseQuery: QueryExp;
  pivotColumns: Array<string>;
  pivotLeafColumn: string | undefined | null;
  baseSchema: reltab.Schema;
  outCols: Array<string>;
  rootQuery: QueryExp | undefined | null;
  sortKey: Array<[string, boolean]>;
  aggMap:
    | {
        [cid: string]: reltab.AggFn;
      }
    | undefined
    | null;

  constructor(
    rt: DataSourceConnection,
    baseQuery: QueryExp,
    baseSchema: reltab.Schema,
    pivotColumns: Array<string>,
    pivotLeafColumn: string | undefined | null,
    outCols: Array<string>,
    rootQuery: QueryExp | undefined | null,
    sortKey: Array<[string, boolean]>,
    inAggMap?: {
      [cid: string]: reltab.AggFn;
    } | null
  ) {
    this.rt = rt;
    this.pivotColumns = pivotColumns;
    this.pivotLeafColumn = pivotLeafColumn;
    this.baseQuery = baseQuery;
    this.baseSchema = baseSchema;
    this.outCols = outCols;
    this.rootQuery = rootQuery;
    this.sortKey = sortKey;
    this.aggMap = inAggMap;
  }
  /*
   * returns a query for the children of the specified path:
   */

  applyPath(path: Path): QueryExp {
    // TODO: Think about how to use rollupBy or a smaller number of groupBys that get chopped up
    // and cache result for efficiency
    // queries are immutable so no need to clone:
    var pathQuery = this.baseQuery; // recCountQuery
    // We will filter by all path components, and then group by the next pivot Column:

    if (path.length > this.pivotColumns.length) {
      throw new Error("applyPath: path length > pivot columns");
    }

    if (path.length > 0) {
      var pred = reltab.and();

      for (var i = 0; i < path.length; i++) {
        let pathElem = path[i];
        let pivotColExp = col(this.pivotColumns[i]);

        if (pathElem == null) {
          pred = pred.isNull(pivotColExp);
        } else {
          pred = pred.eq(pivotColExp, constVal(pathElem));
        }
      }

      pathQuery = pathQuery.filter(pred);
    }

    const pivotColumnInfo: ColumnMapInfo = {
      id: "_pivot",
      displayName: "_pivot",
    };
    const aggCols = this.baseSchema.columns;
    const aggMap = this.aggMap;
    const gbAggs: any =
      aggMap != null ? aggCols.map((cid) => [aggMap[cid], cid]) : aggCols;

    if (path.length < this.pivotColumns.length) {
      /* was: 
      pathQuery = pathQuery
        .groupBy([this.pivotColumns[path.length]], gbAggs)
        .mapColumnsByIndex({
          "0": pivotColumnInfo,
        });
        ...but this leads to ambiguity in the use of the pivot column name, that some SQL engines (BigQuery) don't like,
        so we'll push the definition of _pivot column inside the GroupBy:
      */
      pathQuery = pathQuery
        .extend("_pivot", asString(col(this.pivotColumns[path.length])), {
          displayName: "_pivot",
        })
        .groupBy(["_pivot"], gbAggs);
    } else {
      // leaf level
      const leafExp =
        this.pivotLeafColumn == null
          ? constVal("")
          : asString(col(this.pivotLeafColumn!));
      pathQuery = pathQuery.extend("_pivot", leafExp);
    }

    const depth = path.length + 1;
    pathQuery = pathQuery
      .extend("_depth", constVal(depth))
      .extend("_isRoot", constVal(false))
      .project(this.outCols);
    /*
     * The point of the '_sortVal_<i>' column is that it will be 1 for all rows of
     * depth >= i, 0 for rows where depth < i (which are higher in the pivot tree).
     * We do an ascending sort on this before
     * adding the '_sortVal_<i>_<j>' cols to ensure that parents always come
     * before their children; without this we'd end up putting parent row
     * immediately after children when sorted descending by some column
     */

    const maxDepth = this.pivotColumns.length + 1;

    for (let i = 0; i < maxDepth; i++) {
      const depthVal = depth > i ? 1 : 0;
      pathQuery = pathQuery.extend("_sortVal_" + i, constVal(depthVal));
    }

    for (let i = 0; i < this.pivotColumns.length; i++) {
      let pathElemExp: ValExp = constVal(null);

      if (i < path.length) {
        let colType = this.baseSchema.columnType(this.pivotColumns[i]);
        pathElemExp = constVal(path[i]);
      } else if (i === path.length) {
        pathElemExp = col("_pivot"); // SQL expression referring to _pivot column
      }

      pathQuery = pathQuery.extend("_path" + i, asString(pathElemExp));
    }

    // TODO: Should we optionally also insert _childCount and _leafCount ?
    // _childCount would count next level of groupBy,
    // _leafCount would do count() at point of calculating
    // filter for current path (before doing groupBy).
    // These can certainly have non-trivial costs to calculate

    return pathQuery;
  }
  /*
   * get query for joining with pathQuery to sort to specified depth
   */

  getSortQuery(depth: number): QueryExp {
    let sortQuery = this.baseQuery; // recCountQuery

    const sortCols = this.sortKey.map((p) => p[0]);
    const aggMap = this.aggMap;
    const sortColAggs: any =
      aggMap != null ? sortCols.map((cid) => [aggMap[cid], cid]) : sortCols;
    const gbCols = this.pivotColumns.slice(0, depth);
    sortQuery = sortQuery.groupBy(gbCols, sortColAggs);
    let colMap: { [cid: string]: { id: string } } = {};

    for (let i = 0; i < gbCols.length; i++) {
      const pathColName = "_path" + i;
      colMap[gbCols[i]] = {
        id: pathColName,
      };
    }

    sortQuery = sortQuery.mapColumns(colMap);
    const pathLevel = depth - 1;
    let sortColMap: { [cid: string]: { id: string } } = {};

    for (let i = 0; i < sortCols.length; i++) {
      let colIndex = gbCols.length + i;
      let colName = "_sortVal_" + pathLevel.toString() + "_" + i.toString();
      sortColMap[colIndex.toString()] = {
        id: colName,
      };
    }

    sortQuery = sortQuery.mapColumnsByIndex(sortColMap);
    return sortQuery;
  }
  /*
   * get query for full tree state from a set of openPaths
   */

  getTreeQuery(openPaths: PathTree): QueryExp {
    const maxDepth = this.pivotColumns.length + 1;
    let resQuery = null;

    if (this.rootQuery) {
      resQuery = this.rootQuery;

      for (let i = 0; i < maxDepth; i++) {
        resQuery = resQuery.extend("_sortVal_" + i, constVal(0));
      }

      resQuery = addPathCols(resQuery, 0, maxDepth);
    }

    const openRoot = this.applyPath([]); // immediate children of root

    if (resQuery) {
      resQuery = resQuery.concat(openRoot);
    } else {
      resQuery = openRoot;
    }

    for (let path of openPaths.iter()) {
      let subQuery = this.applyPath(path);
      resQuery = resQuery.concat(subQuery);
    }

    const sortArg: [string, boolean][] = [];

    for (let i = 0; i < maxDepth - 1; i++) {
      sortArg.push(["_path" + i, true]);
    }

    resQuery = resQuery.sort(sortArg);
    return resQuery;
  }
  /*
   * get query for full tree state from a set of openPaths, joined with
   * relevant sort queries based on pivot depth, and with appropriate
   * order by clause
   */

  getSortedTreeQuery(openPaths: PathTree): QueryExp {
    const tq = this.getTreeQuery(openPaths);
    let jtq = tq; // add sort queries for each pivot depth and join to tree query

    for (let i = 0; i < this.pivotColumns.length; i++) {
      let depth = i + 1;
      let sq = this.getSortQuery(depth);

      let joinKey = _.range(0, depth).map((j) => "_path" + j);

      jtq = jtq.join(sq, joinKey);
    }

    // Now let's work out the sort key:
    // potential opt: Eliminate if root not shown

    let tsortKey: [string, boolean][] = [];

    if (this.rootQuery !== null) {
      tsortKey.push(["_isRoot", false]);
    }
    let stq: reltab.QueryExp = jtq;

    if (this.pivotColumns.length > 0 || this.sortKey.length > 0) {
      for (let i = 0; i < this.pivotColumns.length; i++) {
        tsortKey.push(["_sortVal_" + i.toString(), true]);

        // sort keys for this depth:
        let dsortKey: [string, boolean][] = _.range(0, this.sortKey.length).map(
          (j) => ["_sortVal_" + i + "_" + j, this.sortKey[j][1]]
        );

        tsortKey = tsortKey.concat(dsortKey); // splice in path at this depth:

        tsortKey.push(["_path" + i, true]);
      }

      // Add the final _sortVal_i:
      const maxDepth = this.pivotColumns.length;
      tsortKey.push(["_sortVal_" + maxDepth.toString(), true]); // Finally, add the sort key columns itself for leaf level:

      tsortKey = tsortKey.concat(this.sortKey);
    }

    if (tsortKey.length > 0) {
      stq = stq.sort(tsortKey);
    }

    return stq;
  }
}
export const getBaseSchema = (
  rt: DataSourceConnection,
  baseQuery: QueryExp,
  showRecordCount = true
): Promise<Schema> => {
  // add a count column and do the usual SQL where 1=0 trick:
  let schemaQuery = baseQuery.filter(reltab.and().eq(constVal(1), constVal(0)));
  if (showRecordCount) {
    schemaQuery = schemaQuery.extend("Rec", constVal(1));
  }

  const schemap = rt.evalQuery(schemaQuery);
  return schemap.then((schemaRes) => schemaRes.schema);
};
export function vpivot(
  rt: DataSourceConnection,
  baseQuery: QueryExp,
  baseSchema: Schema,
  pivotColumns: Array<string>,
  pivotLeafColumn: string | undefined | null,
  showRoot: boolean,
  sortKey: Array<[string, boolean]>,
  inAggMap:
    | {
        [cid: string]: reltab.AggFn;
      }
    | undefined
    | null = null,
  showRecordCount = true
): VPivotTree {
  const aggMap = inAggMap; // just for Flow

  if (showRecordCount) {
    baseQuery = baseQuery.extend("Rec", constVal(1));
  }
  const hiddenCols = ["_depth", "_pivot", "_isRoot"];
  const outCols = baseSchema.columns.concat(hiddenCols);
  const gbCols = baseSchema.columns.slice();
  const gbAggs: AggColSpec[] =
    aggMap != null ? gbCols.map((cid) => [aggMap[cid], cid]) : gbCols;
  let rootQuery = null;

  if (showRoot) {
    rootQuery = baseQuery
      .groupBy([], gbAggs)
      .extend("_pivot", asString(constVal(null)))
      .extend("_depth", constVal(0))
      .extend("_isRoot", constVal(true))
      .project(outCols);
  }

  return new VPivotTree(
    rt,
    baseQuery,
    baseSchema,
    pivotColumns,
    pivotLeafColumn,
    outCols,
    rootQuery,
    sortKey,
    aggMap
  );
}
