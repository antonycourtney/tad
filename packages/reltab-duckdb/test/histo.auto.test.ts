import * as duckdb from "duckdb-async";
import * as _ from "lodash";
import * as reltab from "reltab";
import {
  asString,
  DataSourceConnection,
  DbDataSource,
  DuckDBDialect,
  tableQuery,
  binsForColumn,
  columnHistogramQuery,
  getNumericColumnHistogramData,
} from "reltab";
import * as reltabDuckDB from "../src/reltab-duckdb";
import * as util from "./testUtils";
import { getFormattedRows } from "./testUtils";

const { col, constVal } = reltab;

const coreTypes = reltab.SQLiteDialect.coreColumnTypes;

let testCtx: DbDataSource;

const q1 = reltab.tableQuery("barttest");

beforeAll(async (): Promise<DataSourceConnection> => {
  const ctx = await reltab.getConnection({
    providerName: "duckdb",
    resourceId: ":memory:",
  });

  testCtx = ctx as DbDataSource;

  const dbds = ctx as DbDataSource;
  const duckDbDriver = dbds.db as reltabDuckDB.DuckDBDriver;

  await reltabDuckDB.nativeCSVImport(
    duckDbDriver.db,
    "test/support/sample.csv"
  );
  await reltabDuckDB.nativeCSVImport(
    duckDbDriver.db,
    "test/support/barttest.csv"
  );

  return testCtx;
});

test("basic column stats and bin count", async () => {
  const qres = await testCtx.evalQuery(q1);

  const statsMap = await testCtx.getColumnStatsMap(q1);

  const tcoeColumnStats = statsMap["TCOE"] as reltab.NumericSummaryStats;

  expect(tcoeColumnStats).toMatchSnapshot();

  const binCount = binsForColumn(tcoeColumnStats);

  console.log("*** number of histogram bins: ", binCount);
});

const intType = DuckDBDialect.columnTypes["INTEGER"];

test("histogram query for column", async () => {
  const schema = await testCtx.getSchema(q1);
  const qres = await testCtx.evalQuery(q1);

  const statsMap = await testCtx.getColumnStatsMap(q1);

  const tcoeColumnStats = statsMap["TCOE"] as reltab.NumericSummaryStats;

  const histoInfo = columnHistogramQuery(q1, "TCOE", intType, tcoeColumnStats);

  console.log("histoInfo: ", histoInfo);

  const dbds = testCtx as DbDataSource;

  console.log("*** histogram query: ", histoInfo!.histoQuery.toJS());

  const sqlQuery = await dbds.getSqlForQuery(histoInfo!.histoQuery);
  console.log("**** histogram sql query: \n", sqlQuery);

  expect(sqlQuery).toMatchSnapshot();

  const histoRes = await testCtx.evalQuery(histoInfo!.histoQuery);
  util.logTable(histoRes);

  const histoData = getNumericColumnHistogramData("TCOE", histoInfo!, histoRes);
  console.log("histogram data:", histoData);
  expect(histoData).toMatchSnapshot();
});

test("full histogram query for all column", async () => {
  const schema = await testCtx.getSchema(q1);

  const statsMap = await testCtx.getColumnStatsMap(q1);

  const [histoInfos, histoQuery] = reltab.getColumnHistogramMapQuery(
    testCtx,
    q1,
    schema,
    statsMap
  );

  console.log("full histogram infos: ", histoInfos);
  const dbds = testCtx as DbDataSource;

  const sqlQuery = await dbds.getSqlForQuery(histoQuery!);
  console.log("full histogram query:\n", sqlQuery);

  expect(sqlQuery).toMatchSnapshot();
});
