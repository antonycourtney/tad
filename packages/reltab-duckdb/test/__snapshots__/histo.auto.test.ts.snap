// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic column stats and bin count 1`] = `
Object {
  "approxUnique": 23,
  "count": 23,
  "max": 419661,
  "min": 104260,
  "pctNull": 0,
  "statsType": "numeric",
}
`;

exports[`full histogram query for all column 1`] = `
"SELECT \\"column\\", \\"bin\\", count(\\"binCount\\") as \\"binCount\\"
FROM (
  SELECT 'Base' as \\"column\\", CAST(floor(((CAST(\\"Base\\" AS DOUBLE) - CAST(50000 AS DOUBLE)) / CAST(42857.142857142855 AS DOUBLE))) AS INTEGER) as \\"bin\\", 1 as \\"binCount\\"
  FROM barttest
)
GROUP BY \\"column\\", \\"bin\\"
UNION ALL
SELECT \\"column\\", \\"bin\\", count(\\"binCount\\") as \\"binCount\\"
FROM (
  SELECT 'TCOE' as \\"column\\", CAST(floor(((CAST(\\"TCOE\\" AS DOUBLE) - CAST(100000 AS DOUBLE)) / CAST(50000 AS DOUBLE))) AS INTEGER) as \\"bin\\", 1 as \\"binCount\\"
  FROM barttest
)
GROUP BY \\"column\\", \\"bin\\"
"
`;

exports[`histogram query for column 1`] = `
"SELECT \\"column\\", \\"bin\\", count(\\"binCount\\") as \\"binCount\\"
FROM (
  SELECT 'TCOE' as \\"column\\", CAST(floor(((CAST(\\"TCOE\\" AS DOUBLE) - CAST(100000 AS DOUBLE)) / CAST(50000 AS DOUBLE))) AS INTEGER) as \\"bin\\", 1 as \\"binCount\\"
  FROM barttest
)
GROUP BY \\"column\\", \\"bin\\"
"
`;

exports[`histogram query for column 2`] = `
Object {
  "binCount": 7,
  "binData": Array [
    10,
    1,
    6,
    3,
    1,
    1,
    1,
  ],
  "binWidth": 50000,
  "brushMaxVal": 500000,
  "brushMinVal": 100000,
  "colId": "TCOE",
  "niceMaxVal": 450000,
  "niceMinVal": 100000,
}
`;
